go/engine/per_user_key_upgrade.go

```diff
@@ -0,0 +1,197 @@
+func (e *PerUserKeyUpgrade) Run(ctx *Context) (err error) {
+	defer e.G().CTrace(ctx.NetContext, "PerUserKeyUpgrade", func() error { return err })()
+	return e.inner(ctx)
+}
+
+func (e *PerUserKeyUpgrade) inner(ctx *Context) error {
+	if !e.G().Env.GetUpgradePerUserKey() {
+		return fmt.Errorf("per-user-key upgrade is disabled")
+	}
+
+	e.G().Log.CDebugf(ctx.NetContext, "PerUserKeyUpgrade load self")
+
+	uid := e.G().GetMyUID()
+	if uid.IsNil() {
+		return libkb.NoUIDError{}
+	}
+
+	loadArg := libkb.NewLoadUserArgBase(e.G()).WithNetContext(ctx.NetContext).WithUID(uid).WithPublicKeyOptional()
+	loadArg.LoginContext = e.args.LoginContext
+	upak, me, err := e.G().GetUPAKLoader().Load(*loadArg)
+	if err != nil {
+		return err
+	}
+
+	e.G().Log.CDebugf(ctx.NetContext, "PerUserKeyUpgrade check for key")
+	if len(upak.Base.PerUserKeys) > 0 {
+		e.G().Log.CDebugf(ctx.NetContext, "PerUserKeyUpgrade already has per-user-key")
+		e.DidNewKey = false
+		return nil
+	}
+	e.G().Log.CDebugf(ctx.NetContext, "PerUserKeyUpgrade has no per-user-key")
+
+	sigKey, encKey, err := e.getDeviceSecretKeys(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Get pukring
+	err = e.G().BumpPerUserKeyring()
+	if err != nil {
+		return err
+	}
+	pukring, err := e.G().GetPerUserKeyring()
+	if err != nil {
+		return err
+	}
+
+	gen1 := keybase1.PerUserKeyGeneration(1)
+
+	pukSeed, err := libkb.GeneratePerUserKeySeed()
+	if err != nil {
+		return err
+	}
+
+	pukReceivers, err := e.getPukReceivers(ctx, upak)
+	if err != nil {
+		return err
+	}
+	if len(pukReceivers) == 0 {
+		return fmt.Errorf("no receivers")
+	}
+
+	// Create boxes of the new per-user-key
+	pukBoxes, err := pukring.PrepareBoxesForDevices(ctx.NetContext,
+		pukSeed, gen1, pukReceivers, *encKey)
+	if err != nil {
+		return err
+	}
+
+	creationTime := e.G().Clock().Now().Unix()
+
+	sig1, err := libkb.PerUserKeyProofReverseSigned(me, pukSeed, gen1, sigKey, creationTime)
+	if err != nil {
+		return err
+	}
+
+	var sigsList []libkb.JSONPayload
+	sigsList = append(sigsList, sig1)
+
+	payload := make(libkb.JSONPayload)
+	payload["sigs"] = sigsList
+
+	libkb.AddPerUserKeyServerArg(payload, gen1, pukBoxes, nil)
+
+	_, err = e.G().API.PostJSON(libkb.APIArg{
+		Endpoint:    "key/multi",
+		SessionType: libkb.APISessionTypeREQUIRED,
+		JSONPayload: payload,
+	})
+	if err != nil {
+		return err
+	}
+	e.DidNewKey = true
+
+	// Add the per-user-key locally
+	err = pukring.AddKey(ctx.NetContext, gen1, pukSeed)
+	if err != nil {
+		return err
+	}
+
+	e.G().UserChanged(uid)
+	return nil
+}
+
+// Get the full keys for this device.
+// Returns (sigKey, encKey, err)
+func (e *PerUserKeyUpgrade) getDeviceSecretKeys(ctx *Context) (libkb.GenericKey, *libkb.NaclDHKeyPair, error) {
+	ad := e.G().ActiveDevice
+	sigKey, err := ad.SigningKey()
+	if err != nil {
+		return nil, nil, err
+	}
+	encKeyGeneric, err := ad.EncryptionKey()
+	if err != nil {
+		return nil, nil, err
+	}
+	encKey, ok := encKeyGeneric.(libkb.NaclDHKeyPair)
```

Comment:
I think in other places we've kept the `GenericKey` type until the lowest possible level, where NaCl functions actually get called. Would doing that simplify the code here? Like right now in team create we just [do this](https://github.com/keybase/client/blob/6f8e1310728baa4b736be141314faf70bf6dcab2/go/engine/team_create.go#L59-L66).